lib/core/src/Cardano/Wallet/DB/Model.hs

```diff
@@ -167,36 +174,85 @@ mRemoveWallet wid db@Database{wallets,txs}
+            , nowPending
+            ]
         }
     )
   where
-    keepBeforePoint :: Wallet s t -> Bool
-    keepBeforePoint cp =
-        (slotId :: BlockHeader -> SlotId) (currentTip cp) <= pt
+    nowPending :: Map (Hash "Tx") TxMeta
+    nowPending = Map.fromList
+        $ fmap (first (txId @t))
+        $ Set.toList
+        $ getPending cp
+
+    wasPending :: TxMeta -> Bool
+    wasPending meta =
+        isPending meta && (slotId :: TxMeta -> SlotId) meta == (tip cp)
 
-    rescheduleOrForget :: TxMeta -> Maybe TxMeta
-    rescheduleOrForget meta = do
-        let isAfter = (slotId :: TxMeta -> SlotId) meta > pt
+mReadCheckpoint :: Ord wid => wid -> ModelOp wid s t xprv (Maybe (Wallet s t))
+mReadCheckpoint wid db@(Database wallets _) =
+    (Right (Map.lookup wid wallets >>= fmap snd . Map.lookupMax . checkpoints), db)
+
+mRollbackTo
+    :: forall s t wid xprv. (Ord wid, DefineTx t)
+    => wid
+    -> SlotId
+    -> ModelOp wid s t xprv ()
+mRollbackTo wid point db@Database{txs} = flip (alterModel wid) db $ \wal ->
+    let
+        nearest = findNearestPoint (Map.elems $ checkpoints wal)
+        txHistory' = Map.mapMaybe (rescheduleOrForget nearest) (txHistory wal)
+        checkpoints' = Map.mapMaybe (keepBeforePoint txHistory') (checkpoints wal)
+    in
+        ((), wal { checkpoints = checkpoints', txHistory = txHistory' })
+  where
+    -- | Removes 'Incoming' transaction beyond the rollback point, and
+    -- reschedule as 'Pending' the 'Outgoing' one beyond the rollback point.
+    rescheduleOrForget :: SlotId -> TxMeta -> Maybe TxMeta
+    rescheduleOrForget nearest meta = do
+        let isAfter = (slotId :: TxMeta -> SlotId) meta > point
         let isIncoming = direction meta == Incoming
         when (isIncoming && isAfter) Nothing
         pure $ if isAfter
-            then meta { slotId = pt, status = Pending }
+            then meta { slotId = nearest, status = Pending }
             else meta
 
+    -- | Find nearest checkpoint's slot before or equal to 'point'
+    findNearestPoint :: [Wallet s t] -> SlotId
+    findNearestPoint = head . sortOn Down . mapMaybe fn
+      where
+        fn :: Wallet s t -> Maybe SlotId
+        fn cp = if (tip cp <= point) then Just (tip cp) else Nothing
+
+    -- | Removes checkpoints beyond the rollback point, and re-calculate the
+    -- pending history for those before the rollback point.
+    keepBeforePoint :: Map (Hash "Tx") TxMeta -> Wallet s t -> Maybe (Wallet s t)
+    keepBeforePoint metas cp = do
+        when (tip cp > point) Nothing
+        pure (updatePending metas cp)
+
+    updatePending :: Map (Hash "Tx") TxMeta -> Wallet s t -> Wallet s t
+    updatePending metas cp = do
+        let pendings = mapMaybe lookupTx $ Map.toList $ Map.filter isPending metas
+        let update tx = newPending @t @s tx . forgetPending (txId @t (fst tx))
+        case reverse $ scanl (flip update) cp pendings of
+            cp':_ -> cp'
+            []    -> cp
```

Comment:
This can maybe be made simpler by reading the pending transactions from the history when we `mReadCheckpoint` maybe :thinking: